toolchain/check/handle_function.cpp

```diff
@@ -55,13 +56,110 @@ static auto BuildFunctionDecl(Context& context, bool is_definition)
       context.node_stack().PopIf<Parse::NodeKind::ImplicitParamList>().value_or(
           SemIR::InstBlockId::Empty);
   auto name_context = context.decl_name_stack().FinishName();
-  auto fn_node =
-      context.node_stack()
-          .PopForSoloParseNode<Parse::NodeKind::FunctionIntroducer>();
+  context.node_stack()
+      .PopAndDiscardSoloParseNode<Parse::NodeKind::FunctionIntroducer>();
+
+  // Process modifiers and introducer.
+  auto first_node = context.innermost_decl().first_node;
+  auto modifiers = ModifiersAllowedOnDecl(context,
+                                          KeywordModifierSet()
+                                              .SetPrivate()
+                                              .SetProtected()
+                                              .SetAbstract()
+                                              .SetDefault()
+                                              .SetFinal()
+                                              .SetImpl()
+                                              .SetVirtual(),
+                                          "`fn` declaration");
+  switch (context.containing_decl().kind) {
+    case DeclState::FileScope:
+      modifiers =
+          ModifiersAllowedOnDecl(context, KeywordModifierSet().SetPrivate(),
+                                 "`fn` declaration at file scope");
+      break;
+
+    case DeclState::Class: {
+      auto access_modifiers = KeywordModifierSet().SetPrivate().SetProtected();
+      modifiers = ModifiersAllowedOnDecl(
+          context, access_modifiers.SetAbstract().SetImpl().SetVirtual(),
+          "`fn` declaration in a `class` definition",
+          context.containing_decl().first_node);
+      if (!context.containing_decl().found.HasAbstract()) {
+        modifiers = ModifiersAllowedOnDecl(
+            context, access_modifiers.SetImpl().SetVirtual(),
+            "`fn` declaration in a non-abstract `class` definition",
+            context.containing_decl().first_node);
+        if (!context.containing_decl().found.HasBase()) {
+          modifiers = ModifiersAllowedOnDecl(
+              context, access_modifiers.SetImpl(),
+              "`fn` declaration in a non-abstract non-base `class` definition",
+              context.containing_decl().first_node);
+        }
+      }
+      break;
+    }
+
+    case DeclState::NamedConstraint:
+      modifiers =
+          ModifiersAllowedOnDecl(context, KeywordModifierSet(),
+                                 "`fn` declaration in a `namespace` definition",
```

Comment:
I think the diagnostics would be clearer if you handled all non-class cases together. That way if I write `impl fn F()` I get told I'm supposed to only write that in a class, not the less-helpful diagnostic that I can't write it in a namespace. That'd also avoid the need to invent diagnostics for impossible situations like an `impl fn` inside a `var` or `let`.

I think restructuring this would also make it easier for me to understand what the code's doing. Eg, with something more like

```
if (kind != DeclScope::Class) {
  modifiers = ModifiersAllowedOnDecl(context, modifiers.UnsetAbstract().UnsetVirtual().UnsetImpl(),
    "`fn` declaration outside of a class");
} else if (!containing_decl_modifiers.HasAbstract() && !containing_decl_modifiers.HasBase()) {
  modifiers = ModifiersAllowedOnDecl(context, modifiers.UnsetVirtual(),
    "`fn` declaration in final class");
} else if (!containing_decl_modifiers.HasAbstract()) {
  modifiers = ModifiersAllowedOnDecl(context, modifiers.UnsetAbstract(),
    "`fn` declaration in non-abstract class");
}
```

... I'd find it more clear which cases permit and disallow abstract / virtual / impl. (Maybe a `ModifiersDisallowedOnDecl` function would make more sense here?)

This restructuring would also allow factoring out handling for subsets of the modifiers into separate functions.